pennylane/templates/subroutines/fable.py

```diff
@@ -166,17 +166,19 @@ def compute_decomposition(input_matrix, wires, tol=0):  # pylint:disable=argumen
                 for c_wire in nots:
                     op_list.append(qml.CNOT(wires=[c_wire] + ancilla))
                 op_list.append(qml.RY(2 * theta, wires=ancilla))
+                nots = {}
                 nots[wire_map[control_index]] = 1
+                continue
+
+            if abs(2 * theta) > tol:
+                for c_wire in nots:
+                    op_list.append(qml.CNOT(wires=[c_wire] + ancilla))
+                op_list.append(qml.RY(2 * theta, wires=ancilla))
+                nots = {}
+            if wire_map[control_index] in nots:
+                del nots[wire_map[control_index]]
             else:
-                if abs(2 * theta) > tol:
-                    for c_wire in nots:
-                        op_list.append(qml.CNOT(wires=[c_wire] + ancilla))
-                    op_list.append(qml.RY(2 * theta, wires=ancilla))
-                    nots = {}
-                if wire_map[control_index] in nots:
-                    del nots[wire_map[control_index]]
-                else:
-                    nots[wire_map[control_index]] = 1
+                nots[wire_map[control_index]] = 1
```

Comment:
Why cannot we set it directly instead of deleting it first? 🤔

Reply:
This check is making sure that CNOTs on the same wire cancel each other out. If `nots[wire_map[control_index]]` is set, we delete it, because it is cancelled out, otherwise we set it to 1 so we can add it to the circuit.